### PR TITLE
issue-697

### PR DESCRIPTION
**Pull request by moley**
Hi, I have fixed the issue, can you please review?
If you have any suggestions please call

### CODE REVIEW

**Comment by donat** [comment]
I'll talk a look today.

**Comment by donat** [comment]
In any case, can you please fix the IP validation failure? Your commit needs a signed-off-by footer with a user that has signed the ECA.

File: org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy [hunk 1]
```diff
@@ -23,6 +24,34 @@ class ImportingWtpProjects extends ProjectSynchronizationSpecification {
     private static final String NON_DEPLOYED = "org.eclipse.jst.component.nondependency"
     private static final String DEPLOYED = "org.eclipse.jst.component.dependency"
     private static final String WTP_COMPONENT_NATURE = "org.eclipse.wst.common.modulecore.ModuleCoreNature"
+	
+	def "Do not check mixed deployment paths without WTP installed"() {
```
Comment (donat) [inline org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy]: ❌ The test fails in my local eclipse as I have WTP installed. I'd fix it the following way:
- Introduce the `isWtpInstalled()` method on the `WorkspaceOperations` interface.
- Adjust `WorkspaceOperationsDelegate` to return a configurable result for `isWtpInstalled`. 
- Duplicate this test to cover both the use cases when WTP is and is not installed.

File: org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy [hunk 2]
```diff
@@ -23,6 +24,34 @@ class ImportingWtpProjects extends ProjectSynchronizationSpecification {
...
+		registerService(UserNotification, notification)
+
+		WorkspaceOperations operations = new WorkspaceOperationsDelegate(recognizeWtpComponentNature: false)
+		registerService(WorkspaceOperations, operations)
+
+		when:
+		importAndWait(root)
+
+		then:
+        0 * notification.errorOccurred(*_)
```
Comment (donat) [inline org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy]: Wrong indentation

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java [hunk 3]
```diff
@@ -42,13 +43,15 @@
     private static final String NON_DEPLOYMENT_ATTRIBUTE = "org.eclipse.jst.component.nondependency";
 
     public static void update(IJavaProject javaProject, OmniEclipseProject project, SubMonitor progress) throws JavaModelException {
+      if (Platform.getBundle("org.eclipse.wst.common.core") != null) {
```
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: ❌ move the check to the `WorkspaceOperations.isWtpInstalled()`  method in order to make it testable.

File: org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy (hunk 2 above)
Comment (Frontrider) [reply org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/workspace/internal/ImportingWtpProjects.groovy]: @moley for this project, set up your formatter to be the same as this repository. If you don't use one, I can recommend it.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java (hunk 3 above)
Comment (moley) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: ok thanks for you suggestions, I will proceed on monday. 
The formatter is not automatically set with oomph? can you please give me a link to the correct formatter?

Comment (Frontrider) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: https://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Freference%2Fpreferences%2Fjava%2Fcodestyle%2Fref-preferences-formatter.htm
https://www.jetbrains.com/help/idea/configuring-code-style.html

Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: Yes, the Buildship Oomph setup contains the formatting configuration.

Comment (moley) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: Sorry - I don't have it yet... The link of @frontrider leads me to the eclipse formatter preferences mechanism - ok. There are three options I could choose (Eclipse [built-in], Eclipse 2.1 [built-in] and Java-Conventions [built-in]. I think none of this tree options is the right one? 
I have set up my workspace with oomph. Which formatter do I have to choose? Is it somewhere commited to git in the projects?

Comment (donat) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: Oomph should set up a new Java formatter called `Buildship`
screen shot 2018-07-24 at 15 27 59
If it doesn't work for you, then just ignore this for this PR. I'll look into the Oomph setup why it's not working for you.

Comment (moley) [reply org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/WtpClasspathUpdater.java]: Sorry, doesn't work, I don't have the buildship formatter in place. I tried to adapt the format manually best possible, don't now if it is 100% exact... So I changed the isWtpInstalled thing and added a second test...
If you have any requests, please call. 
If you need more info to fix the formatter thing, also call. 

Cheers
Markus

**Comment by moley** [comment]
no problem, I have done these commands.

**Comment by donat** [comment]
It seems that only your last commit has the `signed-off-by` footer. Can you please make sure that all commits have that?

**Comment by moley** [comment]
Hi again, I think I've done, please check

**Comment by moley** [comment]
Do you still wait for me to do anything?

**Comment by donat** [comment]
@moley Sorry for the late response, I was on holiday. You don't have to do anything, I'll merge your PR soon.

**Comment by donat** [comment]
@moley Thank you for the contribution!

**Comment by moley** [comment]
@donat No problem, you're welcome. Thank you for merging it, should I delete my branch moley:moley/issue-697?

**Comment by moley** [comment]
@donat 
Hi, 
Did you already build a snapshot, which contains this fix. Can you please give me the url?

**Comment by moley** [comment]
Ok, thank you so far

**Comment by moley** [comment]
@donat 
I have downloaded this versions and the error still occurs :-( Did you check if my change was working on your machine? Did I miss something?

**Comment by moley** [comment]
Different question: I didn't find an uploaded source jar. 
How can I check if my change is contained?

**Comment by Frontrider** [comment]
I woukd like to test the change.

**Comment by donat** [comment]
@moley The change was merged on 7th August, all [snapshot releases](http://download.eclipse.org/buildship/updates/e48/snapshots/2.x/) should contain your change after that.

**Comment by donat** [comment]
Also, we recently [added](https://github.com/eclipse/buildship/issues/621) source reference headers to the jars which contain the commit hash from which the update site was generated. You can use it to trace the history whether your change is included. Or you can use the PDE import feature (referenced at #621), but AFAIC it only works for the Buildship 2.x versions, because the 3.x development is is done on a feature branch.

**Comment by donat** [comment]
@Frontrider You can recreate the sample projects from the [integration tests](https://github.com/eclipse/buildship/pull/720/files#diff-65516a651c52091d6edb472bf08ce11b). Create a project, copy-paste the configuration to your `build.gradle` file and call `Gradle > Refresh Gradle Project`. If everything is right you should not see an exception mentioning mixed WTP deployment paths.

**Comment by moley** [comment]
@donat, @Frontrider Hi, any new insights on this?

**Comment by Frontrider** [comment]
We ended up not using it because the project also has some RCP bits, that will not import at all if we use gradle.

**Comment by moley** [comment]
Hi again @donat 

I retestet my pullrequest and noticed, that the fix was not enough, because it seems that eclipse photon comes with org.eclipse.wst.common_core.feature and org.eclipse.wst.common_ui.feature by default :-( 
So I would suggest to replace Exception by gradle log error or warning. May I open another pullrequest to change implementation like this? Do you have other suggestions to solve this problem?

**Comment by moley** [comment]
Yes, that's the point - as soon as `org.eclipse.wst.common.core` plugin is available, the error is shown. I wanted to avoid this error thinkin that this plugin is not installed by default. But - not using the target platform and downloading the real eclipse - I noticed it is installed. So I think I will try your idea of using release-3.0 branch

Thank you

**Comment by moley** [comment]
Hi again, 

I have tried with branch release-3.0, which did not work. Now I have removed the lines: 
                  ```
  if (deploymentPath != null && !deploymentPath.equals(attribute.getValue())) {
                        throw new UnsupportedConfigurationException("WTP currently does not support mixed deployment paths.");
                    }
```
and created a company internal updatesite, were I installed this version from, which worked. So - what about removing this exception?

**Pull request merged**